src/Components/Browser.JS/src/Boot.Server.ts

```diff
@@ -6,70 +6,97 @@ import { OutOfProcessRenderBatch } from './Rendering/RenderBatch/OutOfProcessRen
+      }
+      catch (ex) {
+        // If there's a rendering exception, notify server *and* throw on client
+        connection.send('OnRenderCompleted', renderId, ex.toString());
+        throw ex;
+      }
+    });
+
+    connection.onclose(error => circuitHandlers.forEach(h => h.onConnectionDown(error)));
+    connection.on('JS.Error', unhandledError.bind(connection));
+
+    window['Blazor'].closeConnection = async() => {
+      await connection.stop();
+      DotNet.attachDispatcher({
+        beginInvokeDotNetFromJS: (...args) => {}});
+    }
 
-  connection.on('JS.BeginInvokeJS', DotNet.jsCallDispatcher.beginInvokeJSFromDotNet);
-  connection.on('JS.RenderBatch', (browserRendererId: number, renderId: number, batchData: Uint8Array) => {
     try {
-      renderBatch(browserRendererId, new OutOfProcessRenderBatch(batchData));
-      connection.send('OnRenderCompleted', renderId, null);
+      await connection.start();
     } catch (ex) {
-      // If there's a rendering exception, notify server *and* throw on client
-      connection.send('OnRenderCompleted', renderId, ex.toString());
-      throw ex;
+      unhandledError.call(connection, ex);
     }
-  });
 
-  connection.on('JS.Error', unhandledError);
+    DotNet.attachDispatcher({
+      beginInvokeDotNetFromJS: (callId, assemblyName, methodIdentifier, dotNetObjectId, argsJson) => {
+        connection.send('BeginInvokeDotNetFromJS', callId ? callId.toString() : null, assemblyName, methodIdentifier, dotNetObjectId || 0, argsJson);
+      }
+    });
 
-  connection.start()
-    .then(async () => {
-      DotNet.attachDispatcher({
-        beginInvokeDotNetFromJS: (callId, assemblyName, methodIdentifier, dotNetObjectId, argsJson) => {
-          connection.send('BeginInvokeDotNetFromJS', callId ? callId.toString() : null, assemblyName, methodIdentifier, dotNetObjectId || 0, argsJson);
-        }
-      });
-
-      // Ensure any embedded resources have been loaded before starting the app
-      await embeddedResourcesPromise;
-
-      connection.send(
-        'StartCircuit',
-        uriHelperFunctions.getLocationHref(),
-        uriHelperFunctions.getBaseURI()
-      );
-    })
-    .catch(unhandledError);
-
-  // Temporary undocumented API to help with https://github.com/aspnet/Blazor/issues/1339
-  // This will be replaced once we implement proper connection management (reconnects, etc.)
-  window['Blazor'].onServerConnectionClose = connection.onclose.bind(connection);
-}
+    return connection;
+  }
 
-function unhandledError(err) {
-  console.error(err);
-
-  // Disconnect on errors.
-  //
-  // TODO: it would be nice to have some kind of experience for what happens when you're
-  // trying to interact with an app that's disconnected.
-  //
-  // Trying to call methods on the connection after its been closed will throw.
-  if (connection) {
-    connection.stop();
+  function unhandledError(this: signalR.HubConnection, err) {
```

Comment:
It's exceptionally weird to use `this` as a regular parameter name. Let's not :)